src/LanguageServer/Impl/TelemetryRpcTraceListener.cs

```diff
@@ -0,0 +1,112 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using Microsoft.PythonTools.Analysis.Infrastructure;
+using StreamJsonRpc;
+using System;
+using System.Diagnostics;
+using System.Reflection;
+
+namespace Microsoft.Python.LanguageServer.Implementation {
+    internal class TelemetryRpcTraceListener : TraceListener {
+        private readonly ITelemetryService2 _telemetryService;
+
+        public TelemetryRpcTraceListener(ITelemetryService2 telemetryService) {
+            _telemetryService = telemetryService;
+        }
+
+        // See JsonRpc.CreateError.
+        public override void TraceData(TraceEventCache eventCache, string source, TraceEventType eventType, int id, params object[] data) {
+            if (eventType != TraceEventType.Error || id != (int)JsonRpc.TraceEvents.LocalInvocationError) {
+                return;
+            }
+
+            Debug.Assert(data.Length >= 2);
+            if (data.Length < 2) {
+                return;
+            }
+
+            var exception = data[0] as Exception;
+            Debug.Assert(exception != null);
+            if (exception == null) {
+                return;
+            }
+
+            // See JsonRpc.StripExceptionToInnerException.
+            if (exception is TargetInvocationException || (exception is AggregateException && exception.InnerException != null)) {
+                exception = exception.InnerException;
+            }
+
+            // Exceptions we expect to throw and pass back over RPC, and therefore should not record.
+            switch (exception) {
+                case InvalidOperationException _:
```

Comment:
In `Rename` I'd change `InvalidOperationException` to `RenameException: NonFatalLanguageServerException` where `NonFatalLanguageServerException: LanguageServerException` and then filter out `NonFatalLanguageServerException`. This way actual bugs that cause `InvalidOperationException` will be logged.

Reply:
I've introduced a new exception and replaced Rename's use of InvalidOperationException with it.